MIPs/mip-3.md

```diff
@@ -0,0 +1,184 @@
+---
+MIP: 3
+Title: Implement `wallet_swapAsset` to support Dapps to swap
+Status: Draft
+Stability: Experimental
+discussions-to: https://github.com/MetaMask/metamask-improvement-proposals/discussions
+Author(s): Tomás Santos (@tommasini)
+Type: Community
+Created: 2023-11-10
+---
+
+## Summary
+
+This proposal introduces a new JSON RPC method `wallet_swapAsset` that allows dApps to request MetaMask to perform a token swap operation. The method accepts parameters for the source and destination tokens and initiates a token swap operation.
+
+## Motivation
+
+This proposal aims to solve a key pain point for dApp developers and users - the inability to easily swap tokens within the dApp interface.
+
+Currently, if a dApp requires users to swap from token A to token B to perform a specific operation, the user has to exit the dApp, go to MetaMask or third parties, search for the token the user needs and execute the swap, then return to the dApp. This disjointed workflow results in a poor user experience.
+
+Implementing `wallet_swapAsset` would allow the entire token swap process to occur seamlessly within the dApp without requiring external integration. This provides the following benefits:
+
+- Improved usability - Users can get the right tokens for dApp operations in fewer steps.
+- Better dApp integration - dApps can build and control the swap experience natively with just a few lines.
+- Increased adoption - More intuitive swap workflow lowers barrier to dApp usage.
+- Interoperability - Standard method works across different wallets.
+- Increase revenue - Easier adoption means more users will use MetaMask swaps.
+
+Comparing with other alternative, implementing it using swaps, would create more friction and less adoption since the users would need to install that specific snap to be able to perform a swap action. The all goal is to reduce friction to swaps.
+
+Overall, this proposal fills a clear gap in dApp capabilities related to swapping tokens. Implementing it would provide significant value to both developers and users in the Ethereum ecosystem.
+
+# Usage Example
+
+```
+ethereum.request({
+  method: 'wallet_swapAsset',
+  params: [{
+    from: [{
+      token_address: '0x1234567890abcdefABCDEF1234567890ABCDEF',
+      amount: '0xDE0B6B3A7640000',
+    }],
+    to: {
+      token_address: '0xabcdef1234567890ABCDEF1234567890abcdef',
+    },
+    user_address: '0x0000000000000000000000000000000000000000'
+  },
+  ]
+});
+
+```
+
+## Proposal Specification
+
+The new JSON RPC method wallet_swapAsset should be implemented with the following parameters:
+
+- `from`: An object containing details about the source token. It should include:
+
+  - `token_address`: The address of the source token.
```

Comment:
Should we specify format here? For example, checksummed address?

Reply:
The openRPC spec may be the best place to specify formats.
Here's doc example with some formatting guidelines included for some params: https://docs.metamask.io/wallet/reference/personal_sign/
And in the api-spec: https://github.com/MetaMask/api-specs/blob/6d55565c9312923b533e580efca88f98364c4da4/openrpc.yaml#L1074